packages/api/src/graphql.js

```diff
@@ -1,111 +1,115 @@
-    const { productIdField, sensitiveFlagField, sensitiveDataField } = this.args;
-    const fields = type.getFields();
-    const hasAuthLogic = !!(productIdField && sensitiveFlagField && sensitiveDataField);
-    const hasAuthFields = !!(
-      Object.keys(fields).includes(productIdField) && 
-      Object.keys(fields).includes(sensitiveFlagField) &&
-      Object.keys(fields).includes(sensitiveDataField)
-    )
-
-    if (!hasAuthLogic) {
-      throw Error('RestrictedProductInformationDirective: missing required arguments');
-    }
-
-    if (!hasAuthFields) {
-      throw Error('RestrictedProductInformationDirective: arguments not in object fields');
-    }
-
-    const originalSensitiveDataFieldResolver = fields[sensitiveDataField].resolve;
-
-    // eslint-disable-next-line max-params
-    fields[sensitiveDataField].resolve = async function(parent, args, context, ...rest) {
-      const productField = parent[productIdField];
-      const isSensitive = parent[sensitiveFlagField];
-      const productRoles = context.me.productRoles;
-      const originalResult = originalSensitiveDataFieldResolver(parent, args, context, ...rest);
-      if (isSensitive) {
-        return productRoles.includes(productField) ? originalResult : null;
-      }
-      return originalResult;
-    }
+class Product {
+	constructor(productData) {
+		Object.assign(this, productData);
+	}
+
+	isRestricted() {
+		return this.isSensitive
+	}
+
+	roleName() {
+		return this.productId
+	}
+}
+
+class SensitiveDataDirective extends SchemaDirectiveVisitor {
+	wrapSensitiveField(field) {
+			if (field._sensitiveWrapped) {
+				return
+			}
+			field._sensitiveWrapped = true
+
+			const { previousResolver = defaultFieldResolver } = field;
+			// eslint-disable-next-line max-params
+			field.resolve = async function(parent, args, context, ...rest) {
+				if (parent.isRestricted()) {
+					const requiredRole = parent.roleName()
+					const userRoles = context.me.roles;
+					if (!userRoles.includes(requiredRole)) {
+						return null;
+					}
+				}
+				return previousResolver(parent, args, context, ...rest);
+			}
+	}
+
+	visitObject(type) {
+		const fields = type.getFields();
+		Object.keys(fields).forEach((fieldName) => {
+			this.wrapSensitiveField(fields[fieldName]);
+		})
+	}
+
+	visitFieldDefinition(field) {
+		this.wrapSensitiveField(field);
   }
 }
 
 
 const schemaText = `
-  directive @restrictedProductInformation(
-    productIdField: String!,
-    sensitiveFlagField: String!,
-    sensitiveDataField: String!,
-  ) on OBJECT
-
-  type SomeValuationQueryResult @restrictedProductInformation(
-    productIdField: "productId",
-    sensitiveFlagField: "isSensitive",
-    sensitiveDataField: "cost",
-  ) {
+	directive @sensitive on OBJECT | FIELD_DEFINITION
+
+  type SomeValuationQueryResult {
     productId: String!
-    cost: Float
+    cost: Float @sensitive
     isSensitive: Boolean
     date: String
   }
 
+	type AnotherQueryResult @sensitive {
```

Comment:
This is just a different type to demonstrate object level directive vs the field level directive above